changefeedccl: implement concurrent execution for webhook sink

spiffyy99 opened the pull request:
Previously, messages sent to the webhook sink were sent synchronously
one after another. This change improves performance by using concurrent
worker goroutines to send sink messages while maintaining ordering
guarantees.

Release note: None

Epic CRDB-7748

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67028)
<!-- Reviewable:end -->

stevendanna reviewed (commented):
Apologies for the review on a draft, but I want to leave a few thoughts before taking off on PTO.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-:-MdT8L4-C9-rxxqTdmej:bjxjgj6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdT4uzeDcMJo4y_BvPZ:-MdT4uzeDcMJo4y_BvP_:b-nh5fhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4903e464ee1ba6c44426efb56c1719c10f284a/pkg/ccl/changefeedccl/sink_webhook.go#L230)):*
> ```Go
> 				err = s.sendMessage(s.ctx, msg)
> 			}
> 			s.workerGroup.GoCtx(func(ctx context.Context) error {
> ```

I'm a little worried about a structure where we spawn a go-routine per message sent.  Further, since it appears we only read from the flush channel when flush is actually being called, we will be accumulating these goroutines for the duration of time between flushes.

One option includes a dedicate result processing goroutine: alongside the worker goroutines, we can start a "result processing" goroutine. This would be responsible just for reading results from the worker goroutines.  Here, we would send our result to that channel directly. While we might block for a moment as it processes the result, that other goroutine should be doing some very quick book-keeping.  This might end up looking similar to what exist in the kafka sync currently.

There are other options as well and I'm not set on the plan above, but I am concerned on not keeping these goroutines around for the time between flushes if we can avoid it.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdT71iZ9nXSibtjgkYg:-MdT71iZ9nXSibtjgkYh:bh6n836) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4903e464ee1ba6c44426efb56c1719c10f284a/pkg/ccl/changefeedccl/sink_webhook.go#L277)):*
> ```Go
> func workerIndex(key []byte) uint32 {
> 	return crc32.ChecksumIEEE(key) % uint32(numWorkers)
> }
> ```

I think future us would benefit from a big block comment explaining the how and why requests are routed to workers.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdT4uzeDcMJo4y_BvPZ:-MdTAvNT1GEsYzCL_t9T:bu0nz2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4903e464ee1ba6c44426efb56c1719c10f284a/pkg/ccl/changefeedccl/sink_webhook.go#L230)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I'm a little worried about a structure where we spawn a go-routine per message sent.  Further, since it appears we only read from the flush channel when flush is actually being called, we will be accumulating these goroutines for the duration of time between flushes.

One option includes a dedicate result processing goroutine: alongside the worker goroutines, we can start a "result processing" goroutine. This would be responsible just for reading results from the worker goroutines.  Here, we would send our result to that channel directly. While we might block for a moment as it processes the result, that other goroutine should be doing some very quick book-keeping.  This might end up looking similar to what exist in the kafka sync currently.

There are other options as well and I'm not set on the plan above, but I am concerned on not keeping these goroutines around for the time between flushes if we can avoid it.

</blockquote></details>

Definitely don't want to spin go routine for each message.
Also heavy +1 on structuring code like in Kafka.
We need something like this because we have add memory monitoring to these inflight messages.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdT4uzeDcMJo4y_BvPZ:-MdTgO_134E_49Ks0tOm:b-atv841) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4903e464ee1ba6c44426efb56c1719c10f284a/pkg/ccl/changefeedccl/sink_webhook.go#L230)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Definitely don't want to spin go routine for each message.
Also heavy +1 on structuring code like in Kafka.
We need something like this because we have add memory monitoring to these inflight messages.

</blockquote></details>

one thing about Kafka though, if we are structuring the code like this, does it even make sense to wrap the inflight logic in its own data structure? seems like we'll need pretty granular control over when we take control of the mutex, so i'm not sure wrapping it offers much benefit with this approach


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdT4uzeDcMJo4y_BvPZ:-MdUDhX67YXM-gWmsMeJ:b-dthljb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4903e464ee1ba6c44426efb56c1719c10f284a/pkg/ccl/changefeedccl/sink_webhook.go#L230)):*
<details><summary><i>Previously, spiffyyeng (Ryan Min) wrote…</i></summary><blockquote>

one thing about Kafka though, if we are structuring the code like this, does it even make sense to wrap the inflight logic in its own data structure? seems like we'll need pretty granular control over when we take control of the mutex, so i'm not sure wrapping it offers much benefit with this approach
</blockquote></details>

Whichever leads to clearer code is fine by me, the wrapping was just an idea to encapsulate some of the logic, but isn't a requirement by any means.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @stevendanna from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwdpY94uW_BNs73aPn:-MdwdpY94uW_BNs73aPo:bhp95zw) (no related file):*
Refactored flushing mechanism to avoid individual goroutines for each message, more like Kafka.

---


<!-- Sent from Reviewable.io -->

miretskiy commented:
@spiffyyeng is this PR still in draft?

spiffyyeng commented:
> @spiffyyeng is this PR still in draft?

@miretskiy nope, we should be good to merge if tests pass. go ahead and review.

spiffyyeng commented:
removing dev-inf as a reviewer for now (change to TC build file was only temporary, will remove once roachtest passes)

spiffyyeng commented on build/teamcity-local-roachtest.sh:
```diff
@@ -34,7 +34,7 @@ build/builder.sh env \
     TC_BUILD_BRANCH="${TC_BUILD_BRANCH-}" \
   stdbuf -oL -eL \
-  ./bin/roachtest run acceptance kv/splits cdc/bank \
+  ./bin/roachtest run cdc/webhook-sink \
```
adding this temporarily to make sure the roachtest passes (will remove afterwards)

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwgx7uDURDjJBp1wTi:-Mdwgx7uDURDjJBp1wTj:b-ova0ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L38)):*
> ```Go
> 	applicationTypeJSON = `application/json`
> 	authorizationHeader = `Authorization`
> 	defaultConnTimeout  = 3 * time.Second
> ```

is this the same default as used by default http client?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwgsjt9nse2S1ZAlin:-Mdwgsjt9nse2S1ZAlio:bqbmszi) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L41)):*
> ```Go
> )
> 
> var numWorkers = workerCount()
> ```

I would drop this constant. 
Instead, I would have a method (just rename workerCount) called defaultWorkerCount().
I would further modify makeWebhookSink to take an explicit parallelism parameter.
Pass this parameter explicitly when constructing webhook sink.  Doing so has few benefits:
1. Makes it explicit at the call site what the parallelism setting is (e.g. `makeWebhookSink(ctx, opts, defaultWorkerCount())
2. Makes it very easy for you to write better test.  In particular, I think you should have a test that exercises parallelism related code.  Running the logic of this sink with numWorkers=1 is important as such setting often uncovers subtle threading bugs.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwhHrG8hKRiipbjPqO:-MdwhHrG8hKRiipbjPqP:b3e7lt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L79)):*
> ```Go
> 	cancelFunc  func()
> 	eventsChans []chan []byte
> 	resultsChan chan error
> ```

we almost certainly would want to have something other than `chan error`
Basically, we need to integrate memory monitoring into this sink.  SO, you will increment memory used when ingesting the message, and you'd decrement it when the message has been emitted.
So, you'd probably need to have something like
```
type webhookResponse struct {
  err error
  size int
}
```

It's okay to just leave a TODO though and add memory monitoring in the follow on pr.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwgXVeCO03_DQvXbKE:-MdwgXVeCO03_DQvXbKF:big70iy) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L155)):*
> ```Go
> 	sink.url = sinkURL{URL: sinkURLParsed}
> 
> 	sink.setupWorkers()
> ```

it actually might make sense to call setupWorkers int he Dial method.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwhwbG63Ft7kKlyHDu:-MdwhwbHEzmQgvmSuKGp:b-nt1whc) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L209)):*
> ```Go
> func workerCount() int {
> 	return system.NumCPU()
> }
> ```

I would just drop this helper -- it's only used to initialize this sink.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 308 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwjg7wDKDhAdc57B9p:-Mdwjg7wDKDhAdc57B9q:b-9lnfsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L308)):*
> ```Go
> // inflightTracker wraps logic for counting number of inflight messages to
> // track when flushing sink
> type inflightTracker struct {
> ```

are you using this struct?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwjioz-VyyJ6tkQZkD:-Mdwjioz-VyyJ6tkQZkE:b8jvpl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L309)):*
> ```Go
> // track when flushing sink
> type inflightTracker struct {
> 	mu struct {
> ```

you probably don't need mu field at all since inflightTracker protects all fields under the lock.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwj_Ri4TCzFZOk8LcD:-Mdwj_Ri4TCzFZOk8LcE:b-svt5py) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L319)):*
> ```Go
> 	ctx context.Context, _ TopicDescriptor, key, value []byte, _ hlc.Timestamp,
> ) error {
> 	index := workerIndex(key)
> ```

nit: I would move workerIndex call directly into select statement.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwkIUFBFzPH6Pn_3u8:-MdwkIUFBFzPH6Pn_3u9:bw1u5sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L368)):*
> ```Go
> 		s.mu.Unlock()
> 		return err
> 	}
> ```

I have a feeling that the flushing logic is not correct.  In particular, when flushing, we must ensure that
all workers complete any messages they may have inflight right now.  it appears that this code synchronizes with the `processResultsLoop`, while what we need is the synchronization with *all* of the `workerLoop`s.

What you probably want is something like this:
```

var wg sync.WaitGroup
wg.Add(numWorkers())
for each worker {
  select {
     .. workerCh <- func() { wg.Done() }
  }
}

wg.Wait()
```

Worker loops needs to be modified to handle a "work group" message.  When received, it calls `wg.Done()`.
It might be beneficial to introduce a wrapper struct around the messages sent to the workers (right now we just send `[]byte`).  Something like:

```
type webhookMessage struct {
  flush func()
  message []byte
}
```
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwn4Y294ryZT82vhoD:-Mdwn4Y3D5R249u5VOmg:bhdvko1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook_test.go#L29)):*
> ```Go
> 	"github.com/stretchr/testify/require"
> )
> ```

let's expand this test to make sure we run test suite with various parallelism settings (numWorkers = 1, 2, 8)


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwkIUFBFzPH6Pn_3u8:-MdwsVm9ApcYC6mYLsTn:bb4v3h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L368)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I have a feeling that the flushing logic is not correct.  In particular, when flushing, we must ensure that
all workers complete any messages they may have inflight right now.  it appears that this code synchronizes with the `processResultsLoop`, while what we need is the synchronization with *all* of the `workerLoop`s.

What you probably want is something like this:
```

var wg sync.WaitGroup
wg.Add(numWorkers())
for each worker {
  select {
     .. workerCh <- func() { wg.Done() }
  }
}

wg.Wait()
```

Worker loops needs to be modified to handle a "work group" message.  When received, it calls `wg.Done()`.
It might be beneficial to introduce a wrapper struct around the messages sent to the workers (right now we just send `[]byte`).  Something like:

```
type webhookMessage struct {
  flush func()
  message []byte
}
```
</blockquote></details>

Sorry, it does appear that this implementation is correct.  Alas, very confusing.  The only reason it works is that webhookSink, itself, is not thread safe, and so you should not have concurrent AddEvent/Flush calls happening
and the synchronization on inflight kinda works.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwsreG2wk-F15CQt3d:-MdwsreG2wk-F15CQt3e:bo7swrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L256)):*
> ```Go
> 			s.mu.inflight--
> 			if s.mu.inflight == 0 && s.mu.flushChan != nil {
> 				s.mu.flushChan <- struct{}{}
> ```

You  should not write to this channel.  You should close it instead.
The reason is that it's possible that we're shutting down, and thhe ctx has already been cancelled.
So, the reader for flushChan (Flush method) already exited.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwkIUFBFzPH6Pn_3u8:-MdwrsEh8m1M-IV_G1bE:bxtfmmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L368)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Sorry, it does appear that this implementation is correct.  Alas, very confusing.  The only reason it works is that webhookSink, itself, is not thread safe, and so you should not have concurrent AddEvent/Flush calls happening
and the synchronization on inflight kinda works.
</blockquote></details>

While I haven't looked in detail at the implementation, I think that in general this structure is correct given the assumption that EmitMessage and Flush are not called concurrently (which they are not right now).

Flush looks like:

- Flush() sets the flushCh under the lock and then blocks on it
- processResultsLoop decrements inflight count for each result it receives
- when inflight count == 0, and flushCh != nil, processResultLoop sends a message on flushCh and unsets flushCh
- Flush() will unblock

When inflight count == 0, we know we don't have any outstanding work because we know EmitMessage isn't being called.

If we wanted to make this safe for concurrent calls to Flush and EmitMessage (the other sinks aren't), we would need to add something that prevented new messages from being enqueued.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @spiffyyeng, and @stevendanna)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwsreG2wk-F15CQt3d:-Mdx0zvrA4c4iYK3LyAa:b-7tj5kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L256)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

You  should not write to this channel.  You should close it instead.
The reason is that it's possible that we're shutting down, and thhe ctx has already been cancelled.
So, the reader for flushChan (Flush method) already exited.

</blockquote></details>

Argh; this is probably correct -- we created flush channel with buffer size of 1.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> Dismissed @miretskiy from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdwgsjt9nse2S1ZAlin:-Mdxefke3yEYkpg6rIsd:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L41)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would drop this constant. 
Instead, I would have a method (just rename workerCount) called defaultWorkerCount().
I would further modify makeWebhookSink to take an explicit parallelism parameter.
Pass this parameter explicitly when constructing webhook sink.  Doing so has few benefits:
1. Makes it explicit at the call site what the parallelism setting is (e.g. `makeWebhookSink(ctx, opts, defaultWorkerCount())
2. Makes it very easy for you to write better test.  In particular, I think you should have a test that exercises parallelism related code.  Running the logic of this sink with numWorkers=1 is important as such setting often uncovers subtle threading bugs.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
updated implementation to use the waitgroup approach discussed internally.

for further context, we've simplified the flush implementation to use a struct (inflightTracker) acting as a wrapper around a WaitGroup, EmitRow() adds one to the WaitGroup and flush simply waits for all the messages to finish sending.

stevendanna commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -220,15 +267,88 @@ func (s *webhookSink) sendMessage(ctx context.Context, reqBody []byte) error {
+	// both the waitgroup and the flushErr at the same time. (i.e. when waiting,
+	// before Done() is called, the error is set, causes deadlock if the same
+	// lock is used)
+	wgMu struct {
```
I think that putting the mutex and the waitgroup in this same structure is a bit confusing.  We should not need a mutex around a waitgroup since the waitgroup is already a concurrency safe datastructure.  However, I believe that this mutex is to prevent Add() calls while we are waiting on Flush()?  If that is correct, let's make the naming of this mutex and the comment reflect that.

stevendanna commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -220,15 +267,88 @@ func (s *webhookSink) sendMessage(ctx context.Context, reqBody []byte) error {
+	defer i.wgMu.Unlock()
+	defer i.errMu.Unlock()
+	i.wgMu.inflightGroup.Wait()
+	// lock here to avoid deadlock with wgMu
```
The deadlock here is that the requests we are Wait'ing on may need to set the error.  If that is correct, let's spell that out explicitly.

stevendanna reviewed (commented):
This definitely feels like it is moving in the right direction.  Each iteration is a bit clearer than the last.

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdy8_UJAc9SzrrbF8nB:-Mdy8_UJAc9SzrrbF8nC:bjyd4kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be57b55efb07a89d85c0698f5a68f9590f4937d/pkg/ccl/changefeedccl/sink_webhook.go#L79)):*
> ```Go
> 	cancelFunc  func()
> 	eventsChans []chan []byte
> 	in          *inflightTracker
> ```

nit: let's rename in with something more descriptive.  inflight for example.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 291 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdy8lun1O5LIo9elsow:-Mdy8lun1O5LIo9elsox:bhzlgeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be57b55efb07a89d85c0698f5a68f9590f4937d/pkg/ccl/changefeedccl/sink_webhook.go#L291)):*
> ```Go
> 	wgMu struct {
> 		syncutil.Mutex
> 		inflightGroup *sync.WaitGroup
> ```

Also, no need to have a pointer receiver for wait group.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 308 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdyA-ge-w7IzllX_uQB:-MdyA-ge-w7IzllX_uQC:b-f2gs6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be57b55efb07a89d85c0698f5a68f9590f4937d/pkg/ccl/changefeedccl/sink_webhook.go#L308)):*
> ```Go
> func (i *inflightTracker) maybeSetError(err error) {
> 	i.errMu.Lock()
> 	defer i.errMu.Unlock()
> ```

I wonder: since we want to save the first error; and the only error, what if instead of using
errMu struct (lock + error), we would just create a `make(chan error, 1)`?

Then: maybeSetError becomes 
```
select {
  case i.errCh <- err:
  default: /* nothing */
}
```

And retrieval of an error is equally simple:

```
select {
  case err := <- i.errCh: 
     return err
  default: /* nothing */
```
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 316 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdy9HfMBE1Hh5ojsFf9:-Mdy9HfMBE1Hh5ojsFfA:bg4r6l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be57b55efb07a89d85c0698f5a68f9590f4937d/pkg/ccl/changefeedccl/sink_webhook.go#L316)):*
> ```Go
> 
> // Add enqueues one inflight message to be flushed.
> func (i *inflightTracker) Add() {
> ```

I would rename this. Maybe to something like `StartRequest`
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 323 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Mdy9N1b1cRrJfc-SH5g:-Mdy9N1b1cRrJfc-SH5h:b-yv65da) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be57b55efb07a89d85c0698f5a68f9590f4937d/pkg/ccl/changefeedccl/sink_webhook.go#L323)):*
> ```Go
> 
> // Done tells the inflight tracker one message has been delivered.
> func (i *inflightTracker) Done() {
> ```

nit: I would rename CompleteRequest (or FinishRequest)


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0tMgl58jMpanfIOqV:-Me0tMgl58jMpanfIOqW:b7hcacc) (no related file):*
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work.  I've left some minor nitpicks and places where we might want some comments/TODO, but overall I like how you've cleared up some of the structure.

---
___
*[pkg/ccl/changefeedccl/sink.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0usg16VTQwHm-x3Ad:-Me0usg16VTQwHm-x3Ae:b-obh571) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbdeb45346b501950403ff02e1b1bcce9a6e8dfc/pkg/ccl/changefeedccl/sink.go#L85)):*
> ```Go
> 			return makeKafkaSink(ctx, sinkURL{URL: u}, feedCfg.Targets, feedCfg.Opts, acc)
> 		case isWebhookSink(u):
> 			return makeWebhookSink(ctx, sinkURL{URL: u}, feedCfg.Opts, defaultWorkerCount())
> ```

Perhaps a TODO to make this configurable somehow.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-MdwkIUFBFzPH6Pn_3u8:-Me0v-Ux3oHS-z480dZM:b-phvivx) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb887e5981cc68ff9a940430616f08edcd4a612c/pkg/ccl/changefeedccl/sink_webhook.go#L368)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

While I haven't looked in detail at the implementation, I think that in general this structure is correct given the assumption that EmitMessage and Flush are not called concurrently (which they are not right now).

Flush looks like:

- Flush() sets the flushCh under the lock and then blocks on it
- processResultsLoop decrements inflight count for each result it receives
- when inflight count == 0, and flushCh != nil, processResultLoop sends a message on flushCh and unsets flushCh
- Flush() will unblock

When inflight count == 0, we know we don't have any outstanding work because we know EmitMessage isn't being called.

If we wanted to make this safe for concurrent calls to Flush and EmitMessage (the other sinks aren't), we would need to add something that prevented new messages from being enqueued.
</blockquote></details>

The waitgroup definitely made this easier to read  I would add a TODO here about the fact that the context is no longer respected.

One solution for that would be do the actual waiting in a goroutine.  The downside of that solution would be that we leak the goroutine in the case of the context cancelling. 
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 326 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0ttBr18fgqap_wGuH:-Me0ttBr18fgqap_wGuI:b-tkdsds) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbdeb45346b501950403ff02e1b1bcce9a6e8dfc/pkg/ccl/changefeedccl/sink_webhook.go#L326)):*
> ```Go
> 
> // Wait waits for all inflight messages to be delivered (inflight = 0) and
> // returns a possible error
> ```

s/error/error./. 

Also, potentially mention that new requests cannot be started until this returns.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 339 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0u2BFCGprxxRqSzkQ:-Me0u2BFCGprxxRqSzkR:b-yrwuv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbdeb45346b501950403ff02e1b1bcce9a6e8dfc/pkg/ccl/changefeedccl/sink_webhook.go#L339)):*
> ```Go
> }
> 
> func (s *webhookSink) EmitRow(
> ```

A doc comment that mentions at least that sink errors won't be seen until Flush() might be good here. Similarly for EmitResolvedTimestampl
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 347 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0u8N08xsoTNIwvOtt:-Me0u8N08xsoTNIwvOtu:b-rk4ozf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbdeb45346b501950403ff02e1b1bcce9a6e8dfc/pkg/ccl/changefeedccl/sink_webhook.go#L347)):*
> ```Go
> 	case <-ctx.Done():
> 		return ctx.Err()
> 	case s.eventsChans[s.workerIndex(key)] <- value:
> ```

Note: We still have a sort of "head of line blocking problem" here if you have multiple updates that map to the same worker.  While we could add a buffer, I think it makes sense to leave that to a follow up, potentially once we have a memory monitor.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 379 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/67028#-Me0uR8x92aooKCtpzqM:-Me0uR8x92aooKCtpzqN:bt0iypj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbdeb45346b501950403ff02e1b1bcce9a6e8dfc/pkg/ccl/changefeedccl/sink_webhook.go#L379)):*
> ```Go
> 	s.cancelFunc()
> 	_ = s.workerGroup.Wait()
> 	for i := 0; i < len(s.eventsChans); i++ {
> ```

nit: You can use range here.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67028)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)



<!-- Sent from Reviewable.io -->

spiffyy99 committed:
changefeedccl: implement concurrent execution for webhook sink

Previously, messages sent to the webhook sink were sent synchronously
one after another. This change improves performance by using concurrent
worker goroutines to send sink messages while maintaining ordering
guarantees.

Resolves #67009

Release note: None

spiffyyeng commented:
roachtest still having issues... going to put it as a todo and merge since it's not urgent

spiffyyeng commented:
bors r+

spiffyyeng commented on pkg/ccl/changefeedccl/sink_webhook.go:
```diff
@@ -220,15 +267,88 @@ func (s *webhookSink) sendMessage(ctx context.Context, reqBody []byte) error {
+	return crc32.ChecksumIEEE(key) % uint32(s.parallelism)
+}
+
+// TODO (ryan min): add memory monitoring for inflight messages
```
added issue tracker: https://github.com/cockroachdb/cockroach/issues/67309

otan commented:
note this has been reverted due to flaking in bors after merging

![image](https://user-images.githubusercontent.com/3646147/125017730-1754c300-e0b7-11eb-9e71-bf9991d3688e.png)

see also: https://github.com/cockroachdb/cockroach/issues/67394
